packages/flutter/lib/src/foundation/consolidate_response.dart

```diff
@@ -3,31 +3,103 @@
 // found in the LICENSE file.
 
 import 'dart:async';
+import 'dart:convert';
 import 'dart:io';
 import 'dart:typed_data';
 
-/// Efficiently converts the response body of an [HttpClientResponse] into a [Uint8List].
+/// Signature for getting notified when chunks of bytes are received while
+/// consolidating the bytes of an [HttpClientResponse] into a [Uint8List].
+///
+/// The `cumulative` parameter will contain the total number of bytes received
+/// thus far. If the response has been gzipped, this number will be the number
+/// of compressed bytes that have received _across the wire_.
+///
+/// The `total` parameter will contain the _expected_ total number of bytes to
+/// be received across the wire (extracted from the value of the
+/// `Content-Length` HTTP response header), or -1 if the size of the response
+/// body is not known in advance.
+///
+/// This is used in [consolidateHttpClientResponseBytes].
+typedef BytesReceivedCallback = void Function(int cumulative, int total);
+
+/// Efficiently converts the response body of an [HttpClientResponse] into a
+/// [Uint8List].
 ///
 /// The future returned will forward all errors emitted by [response].
-Future<Uint8List> consolidateHttpClientResponseBytes(HttpClientResponse response) {
-  // response.contentLength is not trustworthy when GZIP is involved
-  // or other cases where an intermediate transformer has been applied
-  // to the stream.
+///
+/// The [onBytesReceived] callback, if specified, will be invoked for every
+/// chunk of bytes that are received while consolidating the response bytes.
+/// For more information on how to interpret the parameters to the callback,
+/// see the documentation on [BytesReceivedCallback].
+///
+/// If the [response] is gzipped, this will automatically un-compress the
+/// bytes in the returned list (even if [HttpClient.autoUncompress] is false).
+// TODO(tvolkert): Remove the [client] param once https://github.com/dart-lang/sdk/issues/36971 is fixed.
+Future<Uint8List> consolidateHttpClientResponseBytes(
+  HttpClientResponse response, {
+  HttpClient client,
+  BytesReceivedCallback onBytesReceived,
+}) {
   final Completer<Uint8List> completer = Completer<Uint8List>.sync();
-  final List<List<int>> chunks = <List<int>>[];
-  int contentLength = 0;
+
+  final _OutputBuffer output = _OutputBuffer();
+  ByteConversionSink sink = output;
+  int expectedContentLength = response.contentLength;
+  if (response.headers?.value(HttpHeaders.contentEncodingHeader) == 'gzip') {
+    if (client?.autoUncompress ?? true) {
+      // response.contentLength will not match our bytes stream, so we declare
+      // that we don't know the expected content length.
+      expectedContentLength = -1;
+    } else {
+      // We need to un-compress the bytes as they come in.
+      sink = gzip.decoder.startChunkedConversion(output);
```

Comment:
why do we want to override autoUncompress?

Reply:
A search yielded no usages of people calling this method with a client that has `autoUncompress` off, so we have a chance to define the behavior, so I just defined it as "this method will always return uncompressed bytes".  The use case is setting the client to not auto-uncompress so that we can get an accurate (or at least, _more likely to be accurate_) accounting of the expected content length.  When the client has auto-uncompressed, we have no chance of knowing the expected content length.

An alternative would be to add a new `autoUncompress` parameter to this method so callers could control the behavior (it'd be a no-op if the client had already uncompressed).  But since it looked like no one was using this with clients that kept the response bytes compressed, it seemed simpler to just define the method as always uncompressing.

Reply:
How do you download a gzip with this API?

Reply:
heh fair enough - I'll add a parameter 😄